File: c++/objects.cpp

#include <iostream>
#include <vector>
#include <utility>

#define inf 1.0e8

using namespace std;


struct Point{
    float x;
    float y;
};


class Node{
    public:
    struct Point pos;
    float cost;
    Node* parent = NULL;

    Node(float x, float y, float cost=inf){
        pos.x = x;
        pos.y = y;
        cost = cost;
    }
    struct Point getPos(){
        return this->pos;
    }
    float getCost(){
        return this->cost;
    }
    Node* getParent(){
        return this->parent;
    }
    void setCost(float cost){
        this->cost = cost;
    }
    void setParent(Node* p){
        this->parent = p;
    }
};

class Obstacle{
    public:
    vector<Point> vertices;

    Obstacle(vector<Point> v){
        this->vertices = v;
    }
    vector<Point>& getVertices(){
        return this->vertices;
    }
};


class Graph{
    public:
    vector<Node> nodes;

    Node* addNode(Node node){
        this->nodes.push_back(node);
        return &(*(nodes.end()-1));
    }
    vector<Node>* getNodes(){
        // vector<Node*> Ns;
        // for(Node n : this->nodes){
        //     Ns.push_back(&n);
        // }
        return &this->nodes;
    }
};

class ConfigurationSpace{
    public:
    vector<float> xlim;
    vector<float> ylim;
    vector<Obstacle*> obstacles;

    ConfigurationSpace(vector<float> xlm, vector<float> ylm){
        this->xlim = xlm;
        this->ylim = ylm;
    }
    void addObstacle(Obstacle *obs){
        if (this->isWithinLimit(obs)){
            this->obstacles.push_back(obs);
        }
    }
    bool isWithinLimit(Obstacle* obs){
        bool isWithinX = false;
        bool isWithinY = false;
        for(auto xy: obs->getVertices()){
            float x=xy.x;
            float y=xy.y;
            if(xlim[0]<x && x<xlim[1] ){
                isWithinX = true;
            }
            if(ylim[0]<y && y<ylim[1]){
                isWithinY = true;
            }
        }
        if(isWithinX && isWithinY){
            return true;
        }
        return false;
    }
    vector<Obstacle*> getObstacles(){
        return this->obstacles;
    }
};

File: python/objects.py
class Node:
    def __init__(self,x,y,cost=float('inf')):
        self.pos = (x,y)
        self.cost = cost
        self.parent = None

    def getPos(self):
        return self.pos
    
    def getCost(self):
        return self.cost
    
    def setCost(self, cost):
        self.cost = cost
    
    def setParent(self,parent):
        self.parent = parent

    def getParent(self):
        return self.parent

class Graph:
    def __init__(self):
        self.nodes = []
        self.edges = {}
    
    def addNode(self,node):
        self.nodes.append(node)
        self.edges.update({node:[]})
    
    def addEdge(self,n1, n2):
        """
            n1 = first Node
            n2 = second Node

            n1 is parent of n2
        """
        if not n1 in self.nodes:
            self.nodes.append(n1)
            self.edges.update({n1:[]})
        if not n2 in self.nodes:
            self.nodes.append(n2)
            self.edges.update({n2:[]})

        self.edges[n1].append(n2)
        n2.setParent(n1)

    def getNodes(self):
        return self.nodes
    def getEdges(self):
        return self.edges

class Obstacle:
    def __init__(self, vertices):
        """
        vertices -> list of tuples
        """
        self.vertices = vertices
    
    def getVertices(self):
        return self.vertices

class ConfigurationSpace:
    def __init__(self,xlim, ylim, obstacles=[]):
        """
        xlim -> tuple
        ylim -> tuple
        obstacles -> list of "Obstacle" object
        """
        self.xlim = xlim
        self.ylim = ylim
        self.obstacles = obstacles
    
    def addObstacle(self, obstacle):
        if not obstacle in self.obstacles:
            if self.isWithinLimit(obstacle):
                self.obstacles.append(obstacle)

    def isWithinLimit(self,obstacle):
        isWithinX = False
        isWithinY = False
        for xy in obstacle.vertices:
            x,y = xy
            if self.xlim[0] < x < self.xlim[1]:
                isWithinX = True
            if self.ylim[0] < y < self.ylim[1]:
                isWithinY = True
        if isWithinX and isWithinY:
            return True
        return False  

    def getLimit(self):
        return self.xlim, self.ylim

    def getObstacles(self):
        return self.obstacles

